add time_constant function

d-hoshino2626 opened the pull request:
I added some APIs for a time constant measurement with the wire grid and the HWP into the `wiregrid.py`. 
Here is the added functions and its explanation:

1. `_check_wiregrid_position()`
    - This function checks the wire grid is off the window or on the window from the logs for the limit switches.
2. `_stop_hwp_with_wiregrid()`
    - This function confirms the wire grid is on the window and stops the HWP.
3. `_spin_hwp_with_wiregrid(target_hwp_direction)`
    - This function confirms the wire grid is on the window and spins up the HWP in the direction same as `target_hwp_direction`.
    - `target_hwp_direction` is bool argument, `'forward'` or `'backward'`.
4. `_reverse_hwp_with_wiregrid(initial_hwp_direction, streaming=False, stepwise_before=False, stepwise_after=False)`
    - This function reverses the HWP spinning direction from `initial_hwp_direction` using `_stop_hwp_with_wiregrid()` and `_spin_hwp_with_wiregrid(target_hwp_direction)`
    - Also, the stepwise rotation can be performed before and/or after the reversing. The bool arguments `stepwise_before` and `stepwise_after` are to identify whether performing them or not.
    - SMuRF streaming can be performed during the stepwise rotation and HWP speed change.
5. `time_constant(initial_hwp_direction, stepwise_first=True, stepwise_last=True, stepwise_mid=False, repeat=1)`
    - This is the main public API. This function inserts the wire grid, changes the HWP direction as many times as `repeat`, and ejects the wire grid.
    - Also, it takes bias steps before and after the insertion and ejection, execute the stepwise rotations before, in the middle of, or after the HWP direction changes.

Any questions and comments are welcome.

ykyohei reviewed (commented):
I comment in terms of hwp controls for this type of measurements that spins up and down hwp.

1. I suggest streaming separately for forward spin and backward spin. This is because hwp solution pipeline assumes constant rotation direction within the data range. It requires more work to analyze it if both forward and reverse spin are included in the data range.

2. For the hwp rotation direction estimation, you can just use. `apply_hwp_angle_model(tod, on_sign_ambiguous='offcenter')`. It might be possible to simplify functions. The hwp rotation direction estimation has to be different from cmb observations bacause `hwp_solution.pid_direction` is not reliable for spin down measurement. This is because commanded spin direction and actual spin direction is different when hwp is spinning down.` hwp_solution.offcenter_direction` will be the robust rotation direction estimator for this hwp spin down up measurements.

ykyohei commented on src/sorunlib/wiregrid.py:
```diff
@@ -198,6 +198,112 @@ def _check_temperature_sensors():
+        if stepwise_before:
+            rotate(False)
+        # Stop the HWP
+        _stop_hwp_with_wiregrid()
```
I suggest to add following here, to stream separately for forward and backward
```
run.smurf.stream('off')
run.smurf.stream('on')
```

BrianJKoopman replied:
@ykyohei can you take another look at this now that changes have been made. Is this split of streams still required? It looks like a continuous stream while the direction is changed to me at the moment.

ykyohei replied:
Separating streaming is not strictly necessary, but the reason why I recommend it is because it's much more easier to analyze especially when we automate analysis in future.

The hwp angle solution requires correct various effects which can be rotation direction dependent, and current hwp angle solution pipeline assumes that hwp rotation direction is constant within one observation. To apply proper correction on data containing both directions of hwp rotations, we have to either 1. modify the hwp angle solution pipeline or 2. manually split acquired data and re-run the hwp angle solution. 1. requires significant complications to hwp angle solution pipeline and I will not do that. 

My understanding is that wiregrid team preffered manual splitting to avoid any ambiguities of hwp rotation direction. I do not necessarily think that turning off streaming for short period of time leads to ambiguities of hwp rotation direction.

BrianJKoopman replied:
I believe this was resolved in https://github.com/simonsobs/sorunlib/pull/179/commits/43ed0576014b4389aa2a8d9a25044844fdeeee72.

d-hoshino2626 commented:
Thanks Kyohei!
1. I consulted with Hironobu about whether it is better to separate the stream or not and his opinion was not to separate. I heard from him that the HWP pipeline works well if the stream is separated properly by analysis and it is not difficult. He will mainly work on this analysis, so it would be better to follow his opinion, I think.
2.  Thanks for your information! I will re-write it using `apply_hwp_angle_model(tod, on_sign_ambiguous='offcenter')`.

sadachi5 commented on src/sorunlib/wiregrid.py:
```diff
@@ -305,3 +411,80 @@ def calibrate(continuous=False, elevation_check=True, boresight_check=True,
+    run.smurf.bias_step(tag=bs_tag, concurrent=True)
+
+    # Insert the wiregrid
+    insert()
```
Could you add detector streams during the insert() and eject(), and add sleep (5 sec?) before and after the insert()/eject()?
Sleep times have two purposes. One is to see the detector response with/without the wiregrid right before and after the insertion/ejection. The second is to wait for the wiregrid to be stable before the bias-step measurement.

d-hoshino2626 replied:
I added detector streams and sleep. In these streams, I used new tags, `wg_inserting` and `wg_ejecting`. The commit for these modification is here: f1e1157

sadachi5 commented on src/sorunlib/wiregrid.py:
```diff
@@ -305,3 +411,80 @@ def calibrate(continuous=False, elevation_check=True, boresight_check=True,
+    run.smurf.bias_step(tag=bs_tag, concurrent=True)
+
+    # Eject the wiregrid
+    eject()
```
Here, I want to add also detector stream ON --> sleep --> eject --> sleep.

sadachi5 reviewed (commented):
I add comments around insert() and eject().

ykyohei commented:
> 1. I consulted with Hironobu about whether it is better to separate the stream or not and his opinion was not to separate. I heard from him that the HWP pipeline works well if the stream is separated properly by analysis and it is not difficult. He will mainly work on this analysis, so it would be better to follow his opinion, I think.

I'm concerned about future automation of gl calibration. I will let you decide, but I don't see any disadvantage of splitting stream. If you split, you don't need to re-calculate hwp solution.

hnakata-JP commented:
Thanks, Kyohei and Daiki.
As Daiki says, I prefer taking the stream all at once. In my previous analysis, I divided the stream by two by the quad key of hwp_solution and removed about 2000 samples just before/after the split point. What I am concerned about is an opportunity for an interruption between these measurements. I think it would be better to remove any room between them in order not to make any other ambiguous gap.

ykyohei commented:
> Thanks, Kyohei and Daiki. As Daiki says, I prefer taking the stream all at once. In my previous analysis, I divided the stream by two by the quad key of hwp_solution and removed about 2000 samples just before/after the split point. What I am concerned about is an opportunity for an interruption between these measurements. I think it would be better to remove any room between them in order not to make any other ambiguous gap.

OK, I'm fine if you don't mind recalculating the hwp angle.

BrianJKoopman commented on src/sorunlib/wiregrid.py:
```diff
@@ -198,6 +198,112 @@ def _check_temperature_sensors():
+    Args:
+        target_hwp_direction (str): Target HWP direction, 'forward' or 'backward'.
+    """
+    # Check the wiregrid position
+    position = _check_wiregrid_position()
```
The checks in both `_stop_hwp_with_wiregrid()` and `_spin_hwp_with_wiregrid()` seem like are duplicating the check, and as a result making the functions more complicated than they need to be. They're only used in `_reverse_hwp_with_wiregrid()`, which is only called after `insert()`. If `insert()` fails to insert the wiregrid it raises and the schedule with exit.

If you still want to check the wiregrid is inserted, I'd say do the check in `_reverse_hwp_with_wiregrid()` before doing anything else. You can then assume its inserted in these other two functions, and they simplify down to:
```python
def _stop_hwp():
    """Stop the HWP."""
    print("Starting to stop the HWP.")
    run.hwp.stop(active=True)
    print("The HWP stopped successfully.")

```

```python
def _spin_hwp(target_hwp_direction):
    """Spin the HWP to the target direction at 2Hz.
    Args:
        target_hwp_direction (str): Target HWP direction, 'forward' or 'backward'.
    """
    if target_hwp_direction == 'forward':
        print("Starting to spin up the HWP in forward.")
        run.hwp.set_freq(freq=2.0)
        print("The HWP is spinning at 2Hz in forward successfully.")
    elif target_hwp_direction == 'backward':
        print("Starting to spin up the HWP in backward.")
        run.hwp.set_freq(freq=-2.0)
        print("The HWP is spinning at 2Hz in backward successfully.")
```

BrianJKoopman commented on src/sorunlib/wiregrid.py:
```diff
@@ -198,6 +198,112 @@ def _check_temperature_sensors():
+        raise RuntimeError("Initial HWP direction should be either 'forward' or 'backward'. Aborting...")
+
+    current_hwp_direction = initial_hwp_direction
+    if current_hwp_direction == 'forward':
+        target_hwp_direction = 'backward'
+    elif current_hwp_direction == 'backward':
+        target_hwp_direction = 'forward'
```
I want to make sure we get this convention correct. The HWP supervisor agent has an argument called `--forward-dir` that defines 'forward' to be clockwise or counter-clockwise, which seems to imply to me this isn't really a fixed direction. Is there some document about what our convention should be?

(I tried to avoid this question myself when writing `hwp.set_freq()`, but it seems I can't escape it.)

d-hoshino2626 replied:
I found the confluence page about the rotation direction definition: [CHWP main page -- Rotation direction definition](https://simonsobs.atlassian.net/wiki/spaces/PRO/pages/185434251/CHWP+main+page#Rotation-direction-definition). The cw/ccw seems to correspond to the PID value depending on each satp and the PID corresponds to 'forward' or 'reverse' not depending on the satp. We should pay attention which physical direction the 'forward' and 'reverse' mean - cw or ccw. This document uses 'reverse' instead of 'backward'.
Also, I found the 'forwards' and 'backwards' (not 'reverse') in the observation plans on [the confluence](https://simonsobs.atlassian.net/wiki/spaces/SOPS/pages/736591902/Operations+2024-11-02+-+2024-11-08+SATp1). I guess a lot of people may make the 'forwards' correspond to the plus frequency like 'run.hwp.set_freq(freq=+2)`

BrianJKoopman replied:
@sadachi5 I'm still concerned about this, especially after seeing the first link here show different conventions for satp2 compared to the other SATs! There 'forward' is negative and 'backward' is positive. If this is true, this function will end up tagging in correct directions for satp2.

ykyohei replied:
Yes, 'forward' and 'reverse' are not absolute rotation direction. However, the essential points of this calibration is to measure the wiregrid angle as a function of hwp speed, and to flip the rotation direction. Therefore, I think current method is acceptable.

BrianJKoopman replied:
> Yes, 'forward' and 'reverse' are not absolute rotation direction. However, the essential points of this calibration is to measure the wiregrid angle as a function of hwp speed, and to flip the rotation direction. Therefore, I think current method is acceptable.

My concern mostly is with the explicit tagging of "forward" and "reverse" throughout this time constant measurement. If these definitions differ between platforms we'll be generating some incorrect tags.

BrianJKoopman replied:
As you seem to have picked up in 809fc1764437bb44a7d0b9048ce7598e14071e6d, the correct relationship is 'ccw' corresponds to +Hz, and 'cw' to -Hz.

I'm marking this comment as resolved.

BrianJKoopman reviewed (changes requested):
Here's a first pass at a review from me. A couple general comments/comments:
* Please wrap lines to ~80 characters. I'm not super strict on the exact length, but there are some pretty long lines (100+) characters here. Try to be consistent with the existing code here.
* How often will this be run? Daily? Monthly? Once? I was a bit surprised to see the PR, thinking we had implemented the regular calibration in `wiregrid.calibrate()` already.
* We need tests. I've been pretty meticulous to keep 100% coverage from the get-go in this repo. `sorunlib` is tricky because I often can't just interrupt daily operations to test some new functionality. Check out the `tests/` directory to see existing tests. New tests for this will go in `tests/test_wiregrid.py`. There are instructions for running tests in the main README. Also, please merge in the latest `main`, there will be a new code coverage check that runs.

d-hoshino2626 commented:
Thanks Brian! Sorry for the late reply and the lack of explanation. I'm replying about some discussion points.
-  This measurement is to calibrate the detectors' time constant by changing the HWP spin direction with the wiregrid inserted. This will be performed about once a week. Now we change the HWP spin direction once a week, so we will do it on that time.

d-hoshino2626 commented on src/sorunlib/wiregrid.py:
```diff
@@ -198,6 +198,112 @@ def _check_temperature_sensors():
+    """Check the wiregrid position."""
+    actuator = run.CLIENTS['wiregrid']['actuator']
+    resp = actuator.acq.status()
+    ls_status = resp.session['data']['fields']['limitswitch']
+    if (ls_status['LSR1'] == 1 or ls_status['LSL1'] == 1) and not (ls_status['LSR2'] == 1 and ls_status['LSL2'] == 1):
+        position = 'outside'
+    elif (ls_status['LSR2'] == 1 or ls_status['LSL2'] == 1) and not (ls_status['LSR1'] == 1 or ls_status['LSL1'] == 1):
+        position = 'inside'
+    else:
+        raise RuntimeError("The wiregrid position is unknown. The limitswitch response is like this:\n" +
+                           "LSR1: {}\n".format('ON' if ls_status['LSR1'] == 1 else 'OFF')
+                           + "LSL1: {}\n".format('ON' if ls_status['LSL1'] == 1 else 'OFF')
+                           + "LSR2: {}\n".format('ON' if ls_status['LSR2'] == 1 else 'OFF')
+                           + "LSL2: {}\n".format('ON' if ls_status['LSL2'] == 1 else 'OFF')
+                           + "Aborting...")
```
I agree with you. I created the [PR](https://github.com/simonsobs/socs/pull/793) to implement the position data in `session.data`.

BrianJKoopman replied:
Great, I just reviewed that. Once it's merged we should drop the code here and use the `session.data`.

ykyohei commented on src/sorunlib/wiregrid.py:
```diff
@@ -305,3 +411,100 @@ def calibrate(continuous=False, elevation_check=True, boresight_check=True,
+            If this is even, the HWP direction will be the same as the initial direction.
+    """
+
+    # Check the initial HWP direction
+    if initial_hwp_direction not in ['forward', 'backward']:
+        raise RuntimeError("Initial HWP direction should be either 'forward' or 'backward'. Aborting...")
+    current_hwp_direction = initial_hwp_direction
```
'direction' in session.data of pid controller agent might be good for initial direction? This value is not reliable when we spin down hwp, or when hwp is stopped, but it is reliable when hwp is spinning stablly.

d-hoshino2626 replied:
For now, I assumed the scheduler knows the HWP spin direction but this is also under the condition which the HWP is already spinning. Using 'direction' would be more simple than now. I will modify. Thanks Kyohei!

d-hoshino2626 commented:
Hi Brian, I editted the `time_constant` function and some internal functions. I'm sorry for pending so ling time. I'll put the change list here:

1. In `_change_wiregird_position()`, I changed to use `position` of `session.data` in the Wiregrid Actuator Agent.
2. I added `_check_hwp_direction()` function. This function read the `pid_direction` of `session.data` in the HWP PID Agent via the HWP Supervisor Agent and return the current HWP rotation direction, `forward` or `backward`. `0` of `pid_direction` is `forward` and `1` is `backward`.
4. I deleted `_stop_hwp_with_wiregrid()` and `_spin_hwp_with_wiregrid()`. These are integrated `_reverse_hwp_direction()` function.
5. `_reverse_hwp_directoin()` function reverse the HWP in the opposite direction of `initial_hwp_direction` argument. Before and after, the stepwise rotation of wiregrid is executed as necessary. Also, I moved smurf stream parts from here to `time_constant()` function.
6. In `time_constant()` function, I deleted some argments, `initial_hwp_direction`, `stepwise_first`, `stepwise_last`, `stepwise_mid`. The reasons of deleting are here.
    1. `initial_hwp_direction`: Because `_check_hwp_direction()` funcition can check the current hwp rotation direction automatically.
    2. `stepwise_first`, `stepwise_last`, and `stepwise_mid`: Basically, we need to do stepwise rotation before/after the time constant measurement in order to check the data quality. I concluded that skipping stepwise rotations is almost never necessary and I removed these arguments.

Also, I added some tests in `tests/test_wiregrid.py`. The coverage is 100%.

d-hoshino2626 commented:
Hi Brian, I editted the `time_constant` function and some internal functions. I'm sorry for pending so ling time. I'll put the change list here:

1. In `_change_wiregird_position()`, I changed to use `position` of `session.data` in the Wiregrid Actuator Agent.
2. I added `_check_hwp_direction()` function. This function read the `pid_direction` of `session.data` in the HWP PID Agent via the HWP Supervisor Agent and return the current HWP rotation direction, `forward` or `backward`. `0` of pid_direction is `forward` and `1` is `backward`.
3. I deleted `_stop_hwp_with_wiregrid()` and `_spin_hwp_with_wiregrid()`. These are integrated `_reverse_hwp_direction()` function.
4. `_reverse_hwp_directoin()` function reverse the HWP in the opposite direction of `initial_hwp_direction` argument. Before and after, the stepwise rotation of wiregrid is executed as necessary. Also, I moved smurf stream parts from here to `time_constant()` function.
5. In `time_constant()` function, I deleted some argments, `initial_hwp_direction`, `stepwise_first`, `stepwise_last`, `stepwise_mid`. The reasons of deleting are here.
    1. initial_hwp_direction: Because _check_hwp_direction() funcition can check the current hwp rotation direction automatically.
    2. stepwise_first, stepwise_last, and stepwise_mid: Basically, we need to do stepwise rotation before/after the time constant measurement in order to check the data quality. I concluded that skipping stepwise rotations is almost never necessary and I removed these arguments.

Also, I added some tests in tests/test_wiregrid.py. The coverage is 100%.

By the way, I'm leaving the SO collaboration by this month and @sadachi5 is going to take over this PR.

sadachi5 commented:
@BrianJKoopman
Could you check the modification by @d-hoshino2626 ?
We want to merge this new feature for the calibration at SATp1.

BrianJKoopman commented:
> @BrianJKoopman
> Could you check the modification by @d-hoshino2626 ?
> We want to merge this new feature for the calibration at SATp1.

Thanks for the ping, this is on my todo list for tomorrow.

sadachi5 commented on src/sorunlib/wiregrid.py:
```diff
@@ -305,3 +384,116 @@ def calibrate(continuous=False, elevation_check=True, boresight_check=True,
         run.smurf.stream('off')
+
+
+def time_constant(num_repeats=1):
```
Probably, we should `num_repeast=2` not to change the HWP direction in the default value.

BrianJKoopman reviewed (changes requested):
Thanks @d-hoshino2626 for the changes, especially for implementing the tests!

@sadachi5, this is looking good, just a couple of small things.

sadachi5 commented on src/sorunlib/wiregrid.py:
```diff
@@ -413,50 +386,55 @@ def calibrate(continuous=False, elevation_check=True, boresight_check=True,
     # Bias step (the wire grid is off the window)
-    bs_tag = f'biasstep, wg_time_constant, wg_ejected, hwp_2hz_{current_hwp_direction}' + el_tag
+    bs_tag = 'wiregrid, wg_time_constant, wg_ejected, ' + \
+             f'hwp_2hz_{current_hwp_direction}' + el_tag
```
I agree with you. I simply removed "2Hz" in each tags (also for other places).

sadachi5 commented on src/sorunlib/wiregrid.py:
```diff
@@ -465,38 +443,48 @@ def time_constant(initial_hwp_direction, stepwise_first=True, stepwise_last=True
     # Bias step (the wire grid is on the window)
-    bs_tag = f'biasstep, wg_time_constant, wg_inserted, hwp_2hz_{current_hwp_direction}' + el_tag
+    bs_tag = 'wiregrid, wg_time_constant, wg_inserted, ' + \
+             f'hwp_2hz_{current_hwp_direction}' + el_tag
```
Dropped 2hz.

sadachi5 reviewed (approved):
I modified the code by following @BrianJKoopman 's comments. Please check it.

BrianJKoopman reviewed (commented):
Thanks for the changes @sadachi5. I still have some concerns about the rotation direction, see https://github.com/simonsobs/sorunlib/pull/179#discussion_r2047814080.

Also, I'd like some feedback from @ykyohei on his comment in https://github.com/simonsobs/sorunlib/pull/179#discussion_r1805004466.

What's the timeline for the calibration on satp1?

ykyohei reviewed (commented):
Checked the updates and left comments about following points.

1. splitting stream
2. rotation direction
3. bias_step options

sadachi5 commented:
@BrianJKoopman 
I found a critical issue related to the rotation direction. 

The original rotation direction is obtained via ``pid_direction``. This basically checks if the HWP rotation is ``forward`` or ``backward(reverse)``. 
Based on the `pid_direction`, the current ``wiregrid.py`` determins the next frequency (+2Hz or -2Hz) and set the frequency via ``set_freq()`` function. (If the current direction is forward/backward, next freq is -2Hz/+2Hz.)
However, the sign of the frequency (positive or negative) is defined by the direction of counter-clockwise(CCW) or clockwise(CW) seen from the sky. (``CCW`` = +Hz, ``CW`` = -Hz)
As you can see HWP manual [(link)](https://simonsobs.atlassian.net/wiki/spaces/PRO/pages/185434251/CHWP+main+page#Rotation-direction-definition) , 
the relationship between ``CCW``/``CW`` and ``forward"/``backward(reverse)`` can be changed between SATs. 
So, the current procedure to change the HWP direction in ``wiregrid.py`` does not work well.

Thus, I added a function ``get_direction()`` in ``hwp.py``. This needs a variable of ``forward_is_cw``(=``--forward-dir``) defined in the HWP supervisor ocs, but it might be impossible to get its value. If not, do you have any simple idea to get the CCW/CW direction information of the HWP? We need to modify the HWP ocs?

BrianJKoopman commented:
> @BrianJKoopman I found a critical issue related to the HWP rotation direction.
> 
> The original rotation direction is obtained via `pid_direction`. This basically checks if the HWP rotation is `forward` or `backward(reverse)`. Based on the `forward` or `backward`, the current `wiregrid.py` determins the next frequency (+2Hz or -2Hz) and set the frequency via `set_freq()` function. (If the current direction is forward/backward, next freq is -2Hz/+2Hz.) However, the sign of the frequency (positive or negative) is defined by the direction of counter-clockwise(CCW) or clockwise(CW) seen from the sky. (`CCW` = +Hz, `CW` = -Hz) As you can see in the HWP manual [(link)](https://simonsobs.atlassian.net/wiki/spaces/PRO/pages/185434251/CHWP+main+page#Rotation-direction-definition) , the relationship between `CCW`/`CW` and `forward`/`backward(reverse)` depends on SATs. So, the current procedure to change the HWP direction in `wiregrid.py` does not work well.
> 
> Thus, I added a function `get_direction()` in `hwp.py`. This needs a variable of `forward_is_cw`(=`--forward-dir`) defined in the HWP supervisor ocs agent, but it might be impossible to get its value. (See:
> 
> https://github.com/simonsobs/sorunlib/blob/3809f239d378a4efe68e09552e6d151ec9a4e703/src/sorunlib/hwp.py#L70
> )
> If not, do you have any idea to get the CCW/CW direction of the HWP in sorunlib? We need to modify the HWP ocs?

Thanks @sadachi5, that's the same worry I had after seeing that HWP Confluence page in this comment thread: https://github.com/simonsobs/sorunlib/pull/179#discussion_r1811452578

@ykyohei is working on adding the rotation direction to the HWP supervisor agent on the ocs end here: https://github.com/simonsobs/socs/pull/863

And we intend to use that `--forward-dir` argument in the supervisor agent to make sure things are defined consistently: https://github.com/simonsobs/ocs-deployment-configs/pull/436

sadachi5 commented:
@BrianJKoopman 
I largely modified this script on two points:
1. Split the stream between stopping and spinning up the hwp
https://github.com/simonsobs/sorunlib/pull/179#issuecomment-2422445005
2. Change labeling of hwp rotation direction from `forward`/`backward` to `cw`/`ccw`. 
https://github.com/simonsobs/sorunlib/pull/179#discussion_r2058890318

The remaining issue is about the `bias_step` arguments. (I hope this issue will be resolved soon.)
https://github.com/simonsobs/sorunlib/pull/179#discussion_r2047852604 

I think this script is almost ready for implementation.
Please take another look at it.

ykyohei reviewed (commented):
This looks good, thank you for the changes.
Sorry my comment on bias_step was wrong. it's consistent among SATs. bias_dets is different but it's not used here, so it should be ok.

ykyohei commented on src/sorunlib/wiregrid.py:
```diff
@@ -305,3 +384,115 @@ def calibrate(continuous=False, elevation_check=True, boresight_check=True,
+    # Bias step (the wire grid is off the window)
+    bs_tag = 'wiregrid, wg_time_constant, wg_ejected, ' + \
+             f'hwp_{current_hwp_direction}' + el_tag
+    run.smurf.bias_step(tag=bs_tag, concurrent=True)
```
This was wrong, sorry. bias_dets is different, but bias_step is same.

BrianJKoopman reviewed (commented):
> @BrianJKoopman I largely modified this script on two points:
> 
>     1. Split the stream between stopping and spinning up the hwp
>        [add time_constant function #179 (comment)](https://github.com/simonsobs/sorunlib/pull/179#issuecomment-2422445005)
> 
>     2. Change labeling of hwp rotation direction from `forward`/`backward` to `cw`/`ccw`.
>        [add time_constant function #179 (comment)](https://github.com/simonsobs/sorunlib/pull/179#discussion_r2058890318)
> 
> 
> The remaining issue is about the `bias_step` arguments. (I hope this issue will be resolved soon.) [#179 (comment)](https://github.com/simonsobs/sorunlib/pull/179#discussion_r2047852604)
> 
> I think this script is almost ready for implementation. Please take another look at it.

Just taking a quick look I like the change from using 'forward'/'backward' everywhere to using 'cw'/'ccw'. However, I would *really* like to see https://github.com/simonsobs/socs/pull/863 settle and get merged before giving my final review here, as that has direct impact on the relationship between +/- frequency and 'cw'/'ccw' direction.

sadachi5 commented on src/sorunlib/hwp.py:
```diff
@@ -43,3 +43,24 @@ def stop(active=True, brake_voltage=None):
+    if direction not in ['cw', 'ccw']:
+        raise RuntimeError("The HWP direction is unknown. Aborting...")
+
+    return direction
```
Done

sadachi5 reviewed (commented):
Rename: hwp.get_direction() --> hwp._get_direction()

BrianJKoopman reviewed (approved):
Alright, I think this is as ready as it will be. I'm still not a fan of this hardcoded correspondance between direction and the sign of the frequency:
```
if target_hwp_direction == 'ccw':
    run.hwp.set_freq(freq=2.0)
elif target_hwp_direction == 'cw':
    run.hwp.set_freq(freq=-2.0)
current_hwp_direction = target_hwp_direction
```

This won't improve until the agent is modified though, so I added a note in the comments here for future reference. I think it would be worth removing this, though I expect in the near term the likelihood that this correspondence changes will be low.

Thanks for all the hard work, tracking of different HWP changes, and updates on this one! @sadachi5 @d-hoshino2626 @ykyohei
